### ISSUE TITLE
a11y bug: number filter should insert a minus character for negative values (not dash)

### ISSUE THREAD

**Issue by 120psi** [angular/angular.js#15550]
I noticed that the AngularJS number formatter adds a dash for negative numbers (defined in ngLocale) but, for a11y, we should be using the minus sign (utf8 "−") (entity &minus;) which is read out by screen readers as "minus" whereas a dash is often omitted.

**Comment by naomiblack** [comment]
@petebacondarwin assigning to you for triage... I'm not sure who should pick this one up.

**Comment by gkalpak** [comment]
**Adding some context:**

Angular uses the [Google Closure I18N library][1], to generate its own I18N files. For negative numbers in particular, we rely on the negative patterns (if present) found in [numberformatsymbols.js][2]. If there is no negative pattern (the majority of cases), we [indeed use][3] `-` (a dash) - we also have a comment saying it is the same as `\u2212` (the minus sign), which is obviously not very accurate.

I would be preferable (imo) to use the locale's [`MINUS_SIGN`][4] instead. That said, this would still not solve the issue by itself, since most locales use a dash as their `MINUS_SIGN`. There are also some that use `\u2013` (an [n-dash][5]) as their `MINUS_SIGN` ([example][6]) and I assume there are some that use `\u2212`.

##
Ideally, we should be relying on Closure's `MINUS_SIGN` and the `MINUS_SIGN`s should be fixed in Closure, but I am not sure if there are legitimate reasons for the current values in Closure.

In any case, I don't think this is something that the number formatter should overwrite. It should stick to the `negPre`/`negSuf` values specified in each locale. Any solution should go through the locale files.

Alternatively, we could have `ngAria` overwrite the dash in the current locale's `negPre`/`negSuf`, although that is the kind of magic that turns out to be more confusing than helpful.

Finally, as a temporary workaround, it is also possible to overwrite the minus signs for number/currency formatting ([demo][7]):

```js
run(function ($locale) {
  // `number` patterns
  replaceMinusSign($locale.NUMBER_FORMATS.PATTERNS[0]);

  // `currency` patterns
  replaceMinusSign($locale.NUMBER_FORMATS.PATTERNS[1]);

  function replaceMinusSign(patternsObj) {
    patternsObj.negPre = patternsObj.negPre.replace(/-/g, '\u2212');
    patternsObj.negSuf = patternsObj.negSuf.replace(/-/g, '\u2212');
  }
})
```

##
And with that, I hand it back to @petebacondarwin :grin:

[1]: https://github.com/google/closure-library/tree/master/closure/goog/i18n
[2]: https://github.com/google/closure-library/blob/3589a5c7d0236c84d6f2c5d01b036e782af41acd/closure/goog/i18n/numberformatsymbols.js
[3]: https://github.com/angular/angular.js/blob/a24777a2c4ad2ac087d9e3aa278fa2e61e8cc740/i18n/src/parser.js#L60-L61
[4]: https://github.com/google/closure-library/blob/3589a5c7d0236c84d6f2c5d01b036e782af41acd/closure/goog/i18n/numberformatsymbols.js#L165
[5]: http://www.fileformat.info/info/unicode/char/2013/index.htm
[6]: https://github.com/google/closure-library/blob/3589a5c7d0236c84d6f2c5d01b036e782af41acd/closure/goog/i18n/numberformatsymbols.js#L2148
[7]: https://codepen.io/ExpertSystem/pen/eBqWza?editors=0010

**Comment by IgorMinar** [comment]
Let's get this fixed in [CLDR](http://cldr.unicode.org/) itself as the problem seems to originate there. Check out: http://www.unicode.org/cldr/charts/latest/summary/en.html which uses `\u002d` for the minus sign, rather than the accessible version: `\u2212`.

Once it is fixed there, we should ensure that the fix correctly propagates to Closure but also to angular/angular which pulls locale info directly from CLDR (fyi: @vicb, @pkozlowski-opensource, @ocombe).

**Comment by IgorMinar** [comment]
I've ping the google i18n team to help us determine the next steps to get this fixed.

**Comment by 120psi** [comment]
Original reporter here.  This actually may be screen-reader dependent and there might not be a straightforward solution.  I have access to both ChromeVox and VoiceOver.

Further investigation:

* VoiceOver reads '-13' aloud as "13"
* VoiceOver reads '&minus;13' as "minus 13"
* ChromeVox reads '-13' aloud as "minus 13"
* ChromeVox reads '&minus;13' as "minus sign 13"

I'd want to find out what JAWS and NVDA do before moving forward, since those are the most commonly used readers in the real world.

**Comment by petebacondarwin** [comment]
Waiting on further information from @120psi

**Comment by NilsGoogle** [comment]
(replying from corp github account)

Attached is a simple HTML page with both &minus; and - used.
[test.html.txt](https://github.com/angular/angular.js/files/755743/test.html.txt)

As stated earlier in the thread, this appears to be an implementation issue with VoiceOver vs. a need to fix underlying libraries

**Comment by petebacondarwin** [comment]
So does that mean that we can close this issue @NilsAtGoogle ?

**Comment by NilsAtGoogle** [comment]
I can't test with JAWS or NVDA and would prefer someone to take a look with those screen readers, but I think it's fine to close this out.  Thanks for helping out.

I'd like to think that screenreaders should adhere to common existing usage of characters (i.e. it's much easier to fix a screenreader than fix all negative numbers on the internet) so perhaps bugs should be filed against the screenreader makers where applicable.

**Comment by Narretz** [comment]
@marcysutton do you maybe have access to JAWS / NVDA and could test what they make of the negative number examples in https://github.com/angular/angular.js/issues/15550#issuecomment-277799518?

**Comment by jweir** [comment]
Something to consider when not using the hyphen-minus – the number may no longer be portable.  When using the unicode minus character, copying and pasting negative numbers into applications (such as Excel, a Javascript console, etc) may result in unexpected behavior.

**Comment by thedamon** [comment]
Someone in a similar but not directly related discussion on the W3C mailing list captured [minus behaviour of various screenreaders here](https://docs.google.com/spreadsheets/d/1FwB3eVebNThYlY0mo1ZeKui3QunmgR693_h8DfLH0OQ/edit).

They seem to conclude the minus symbol is the best overall choice, but I'm concerned by the copy/pasting unpredictabilities mentioned by @jweir ... Do we know which applications do or don't have a problem with that?
